test/python/visualization/test_circuit_text_drawer.py

```diff
@@ -5208,7 +5208,7 @@ def test_open_controlled_u1(self):
         expected = "\n".join(
             [
                 "                                                           ",
-                "qr_0: |0>─o─────────o─────────o─────────o─────────■────────",
+                "qr_0: |0>─■─────────o─────────o─────────o─────────■────────",
```

Comment:
I am not sure why this test changed, but I don't think it's wrong like this.

Reply:
I think you'll need to update the logic here: https://github.com/Qiskit/qiskit/blob/main/crates/circuit/src/circuit_instruction.rs#L834-L851 or in the circuit method (it depends how the gate is being appended to the circuit).

The issue is a controlled gate with a non default ctrl_state is not a standard gate. The standard gate definition is not taken into account and we should treat a custom `ctrl_state` as a `PyGate`. It looks like the logic I had for this only covers when the input is a singleton which ignores standard controlled gates that take a parameter.

Reply:
Updated logic and added test in [9b47121](https://github.com/Qiskit/qiskit/pull/12659/commits/9b47121e1afb51e8f3bdd609edbb67939fd0286d)